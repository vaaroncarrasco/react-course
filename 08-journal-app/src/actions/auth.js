import Swal from "sweetalert2";

import { firebase, googleAuthProvider } from "../firebase/firebase-config";
import { types } from "../types/types"
import { noteLogout } from "./notes";
import { finishLoading, startLoading } from "./ui";

export const startLoginEmailPassword = (email, password) => {
    return (dispatch) => {

        dispatch( startLoading() );

        return firebase.auth().signInWithEmailAndPassword(email, password)
            .then( ( { user } ) => {

                dispatch(
                    login( user.uid, user.displayName )
                )

                dispatch( finishLoading() );


            }).catch( err => {
                dispatch( finishLoading() );
                console.log(err);
                Swal.fire('Error', err.message, 'error');
            });

    }
}


export const startRegisterWithEmailPasswordName = ( email, password, name ) => {

    // ? async fns return (callbacks) => {}
    return ( dispatch ) => {

        firebase.auth().createUserWithEmailAndPassword( email, password )
            .then( async ( {user} ) => {

                await user.updateProfile({ displayName: name });

                console.log(user)


                dispatch(
                    login( user.uid, user.displayName )
                )

            }).catch( err => {
                console.log(err);
                Swal.fire('Error', err.message, 'error');
            });

    }

}


export const startGoogleLogin = () => {

    return (dispatch) => {

        firebase.auth().signInWithPopup( googleAuthProvider )
            .then( ( { user } ) => {
                dispatch(
                    login( user.uid, user.displayName )
                )
            });

    }

}

export const login = (uid, displayName) => ({
    type: types.login,
    payload: {
        uid,
        displayName
    }
})

export const startLogout = () => {

    // ? Returns an async callback

    return async (dispatch) => {
        await firebase.auth().signOut();

        dispatch( logout() );

        dispatch( noteLogout() );
    }
}

export const logout = () => ({
    type: types.logout
})