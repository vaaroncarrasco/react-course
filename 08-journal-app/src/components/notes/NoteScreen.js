import React, { useEffect, useRef } from 'react'
import { useDispatch } from 'react-redux';
import { useSelector } from 'react-redux'
import { activeNote, startDeleting } from '../../actions/notes';
import { useForm } from '../../hooks/useForm';
import { NotesAppBar } from './NotesAppBar'

export const NoteScreen = () => {

    const dispatch = useDispatch();

    // ? useForm -> for note input values and fill them if they already exist
    const { active: note } = useSelector(state => state.notes);
    const [ formValues, handleInputChange, reset ] = useForm( note );
    const { body, title, id } = formValues;

    // ? useRef() react hook -> stores mutable var without re running comp if changes
    const activeId = useRef( note.id ); // {}.current -> current value
    useEffect(() => { // be aware of infinite loops
        if ( note.id !== activeId.current ) {
            reset( note ); // set note new state
            activeId.current = note.id;
        }
    }, [note, reset]);

    useEffect(() => {

        dispatch( activeNote( formValues.id, { ...formValues } ) );

    }, [formValues, dispatch]);

    const handleDelete = () => {
        dispatch( startDeleting( id ) );
    }


    // ? ALWAYS REMEMBER TO ADD name, value, onChange -> input properties
    // for correct form + useState usage

    return (
        <div className="notes__main-content">

            <NotesAppBar />

            <div className="notes__content">

                <input
                    type="text"
                    placeholder="Some awesome title"
                    className="notes__title-input"
                    autoComplete="off"
                    name="title"
                    value={ title }
                    onChange={ handleInputChange }
                />

                <textarea
                    placeholder="What happened today"
                    className="notes__textarea"
                    name="body"
                    value={ body }
                    onChange={ handleInputChange }
                ></textarea>

                {
                    (note.url) &&
                        <div className="notes__image">
                            <img
                                src={ note.url }
                                alt="imagen"
                            />
                        </div>
                }

            </div>

            <button
                className="btn btn-danger"
                onClick={ handleDelete }
            >
                Delete
            </button>

        </div>
    )
}
