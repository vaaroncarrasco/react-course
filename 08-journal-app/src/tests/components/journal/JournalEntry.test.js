import React from 'react'
import '@testing-library/jest-dom';
import { mount } from 'enzyme';
import { Provider } from "react-redux";
import thunk from 'redux-thunk';
import configureStore from 'redux-mock-store' //ES6 modules

import { JournalEntry } from '../../../components/journal/JournalEntry';
import { activeNote } from '../../../actions/notes';

const middlewares = [ thunk ];
const mockStore = configureStore(middlewares);

const initState = {};

let store = mockStore(initState);
store.dispatch = jest.fn();

const nota = {
    id: 1,
    date: 0,
    title: 'hola',
    body: 'mundo',
    url: 'https://domain.com/foto.jpg'
}

const wrapper = mount(
    <Provider store={ store }>
        <JournalEntry {...nota} />
    </Provider>
);


describe('testing <JournalEntry />', () => {

    test('should match snapshot', () => {
        expect( wrapper ).toMatchSnapshot();
    });

    test('should call activeNote', () => {

        wrapper.find('.journal__entry').prop('onClick')();

        expect( store.dispatch ).toHaveBeenCalledWith(
            activeNote( nota.id, { ...nota } )
        );

    });

});