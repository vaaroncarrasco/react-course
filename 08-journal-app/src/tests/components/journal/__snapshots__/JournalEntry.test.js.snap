// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing <JournalEntry /> should match snapshot 1`] = `
<div
  className="journal__entry pointer"
  onClick={[Function]}
>
  <div
    className="journal__entry-picture"
    style={
      Object {
        "backgroundImage": "url(https://domain.com/foto.jpg)",
        "backgroundSize": "cover",
      }
    }
  />
  <div
    className="journal__entry-body"
  >
    <p
      className="journal__entry-title"
    >
      hola
    </p>
    <p
      className="journal__entry-content"
    >
      mundo
    </p>
  </div>
  <div
    className="journal__entry-date-box"
  >
    <span>
      Wednesday
    </span>
    <h4>
      31st
    </h4>
  </div>
</div>
`;
